### PR TITLE
feat(APIM-79): add endpoint to get parties by search text

### PR DESCRIPTION
**Pull request by oscar-richardson-softwire**
## Introduction
This PR is to add the `GET /party?searchText={searchText}` endpoint.

## Resolution
The endpoint:

- gets an ID token from FIS IdP,
- sends an authenticated `GET /Party/Search/{searchText}` to ACBS,
- transforms the response to extract the relevant fields.

To do this, I've added:
- a `party` module for the new controller
- a `get-party-by-search-text` service in the `party` module

### CODE REVIEW

**Comment by oscar-richardson-softwire** [comment]
Should this endpoint be renamed to `/parties` in accordance with API spec guidelines, as it returns an array of parties that match the search text?

**Comment by oscar-richardson-softwire** [comment]
Should this endpoint use `searchtext` (as Mulesoft does) or `searchText` (in line with the camel casing used for other parameters in other endpoints, e.g. `partyIdentifier`)?

File: src/modules/party/dto/acbs-get-party-by-search-text-response-element.dto.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export interface AcbsGetPartyBySearchTextResponseElement {
+  PartyAlternateIdentifier: string;
+  IndustryClassification: { IndustryClassificationCode: string };
+  PartyName1: string;
+  PartyName2: string;
+  PartyName3: string;
+  MinorityClass: { MinorityClassCode: string };
+  CitizenshipClass: { CitizenshipClassCode: string };
+  OfficerRiskDate: string;
```
Comment (lmarrai-sw) [inline src/modules/party/dto/acbs-get-party-by-search-text-response-element.dto.ts]: We discussed whether or not we should have `axios` parse the datetime string here. We decided not to do it yet but we might change it in future. We should raise a ticket about this
(I'll take that action and resolve this comment when done)

File: src/modules/party/dto/party-query.dto.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsNotEmpty, Length } from 'class-validator';
+
+export class PartyQueryDto {
+  @IsNotEmpty()
+  @Length(3)
+  @ApiProperty()
+  public searchText: string;
```
Comment (lmarrai-sw) [inline src/modules/party/dto/party-query.dto.ts]: There's no need to specify `public` here, since `public` is the default

File: src/modules/party/dto/party-query.dto.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsNotEmpty, Length } from 'class-validator';
+
+export class PartyQueryDto {
+  @IsNotEmpty()
+  @Length(3)
```
Comment (lmarrai-sw) [inline src/modules/party/dto/party-query.dto.ts]: I didn't think about this before, but is there any point specifying `IsNotEmpty()` and `Length(3)` at the same time? Does `Length(3)` cover all the cases of `IsNotEmpty()` anyway?

File: src/modules/party/dto/party-query.dto.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+  @ApiProperty()
```
Comment (lmarrai-sw) [inline src/modules/party/dto/party-query.dto.ts]: The swagger docs we're generating don't seem to know that `searchText` has to be at least 3 characters - can we find a way to add that info to the swagger docs please?

File: test/party/get-party-by-search-text.api-test.ts [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+  };
+
+  const givenAuthenticationWithTheIdpSucceeds = (): void => {
+    givenCreatingASessionWithTheIdpSucceeds();
+    givenGettingAnIdTokenFromTheIdpSucceeds();
+  };
+
+  const requestToGetPartyBySearchText = (): nock.Interceptor =>
+    nock('https://test-acbs-url.com/base-url').get(`/Party/Search/${searchText}`).matchHeader('authorization', `Bearer ${idToken}`);
+});
```
Comment (lmarrai-sw) [inline test/party/get-party-by-search-text.api-test.ts]: Watch out for when you next rebase onto a later version of APIM-76, the config values for ACBS (like `https://test-acbs-authentication-url.com/base-url`) will be different so this part of your code will need to be updated. You will be able to see how the API test in APIM-76 does it

File: test/party/get-party-by-search-text.api-test.ts [hunk 6]
```diff
@@ -0,0 +1,284 @@
+import { Api } from '@ukef-test/support/api';
+import { RandomValueGenerator } from '@ukef-test/support/random-value-generator';
+import nock from 'nock';
+
+describe('GET /party?searchText={searchText}', () => {
+  const valueGenerator = new RandomValueGenerator();
+
+  const searchText = valueGenerator.stringOfNumericCharacters();
```
Comment (lmarrai-sw) [inline test/party/get-party-by-search-text.api-test.ts]: We should make sure `searchText` has at least 3 characters here - we'll need to edit the `stringOfNumericCharacters` function to allow you to specify a minimum length

File: test/party/get-party-by-search-text.api-test.ts [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+  it('returns a 400 response if searchText is less than 3 characters', async () => {
+    const { status, body } = await api.get(`/api/v1/party?searchText=00`);
+
+    expect(status).toBe(400);
+    expect(body).toStrictEqual({
+      statusCode: 400,
+      error: 'Bad Request',
+      message: ['searchText must be longer than or equal to 3 characters'],
+    });
+  });
```
Comment (lmarrai-sw) [inline test/party/get-party-by-search-text.api-test.ts]: Can we test the case where searchText has exactly 3 characters as well please?

File: src/modules/party/get-party-by-search-text.service.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+      .then((acbsResponse) =>
+        acbsResponse.data.map((element) => ({
+          alternateIdentifier: element.PartyAlternateIdentifier,
+          industryClassification: element.IndustryClassification.IndustryClassificationCode,
+          name1: element.PartyName1,
+          name2: element.PartyName2,
+          name3: element.PartyName3,
+          smeType: element.MinorityClass.MinorityClassCode,
+          citizenshipClass: element.CitizenshipClass.CitizenshipClassCode,
+          officerRiskDate: element.OfficerRiskDate.slice(0, 10),
```
Comment (lmarrai-sw) [inline src/modules/party/get-party-by-search-text.service.ts]: `OfficerRiskDate` can be `null` unfortunately - we have to check that before we slice

File: src/modules/party/get-party-by-search-text.service.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+    const acbsRequest = {
+      baseURL: this.config.baseUrl,
+      headers: {
+        Authorization: 'Bearer ' + token,
+      },
+    };
+
+    const response = await lastValueFrom(
+      this.httpService.get<AcbsGetPartyBySearchTextResponseElement[]>(GetPartyBySearchTextService.path + '/' + searchText, acbsRequest),
```
Comment (lmarrai-sw) [inline src/modules/party/get-party-by-search-text.service.ts]: Can we use string interpolation instead of string concatenation please?
```
`${GetPartyBySearchTextService.path}/${searchText}`
```

File: src/modules/party/party.controller.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  })
+  @ApiOkResponse({
+    description: 'The matching parties have been successfully retrieved.',
+    type: GetPartyBySearchTextResponseElement,
+    isArray: true,
+  })
+  @ApiInternalServerErrorResponse({
+    description: 'An internal server error has occurred.',
+  })
+  async getPartyBySearchText(@Query() query: PartyQueryDto): Promise<GetPartyBySearchTextResponse> {
```
Comment (lmarrai-sw) [inline src/modules/party/party.controller.ts]: Slight nitpick: this would probably be a bit clearer as `getPartiesBySearchText` or `findPartiesBySearchText` or `searchParties` - or something along those lines

I have a similar comment for the method on the service

File: src/modules/party/dto/get-party-by-search-text-response-element.dto.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
+import { ApiResponseProperty } from '@nestjs/swagger';
+
+export type GetPartyBySearchTextResponse = GetPartyBySearchTextResponseElement[];
+
+export class GetPartyBySearchTextResponseElement {
+  @ApiResponseProperty()
+  alternateIdentifier: string;
```
Comment (IgnatG) [inline src/modules/party/dto/get-party-by-search-text-response-element.dto.ts]: are any of these properties required? i.e. do we need to check whether they exist or not? it's the same for all other properties below

File: src/modules/party/get-party-by-search-text.service.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
+import { HttpService } from '@nestjs/axios';
+import { Inject, Injectable } from '@nestjs/common';
+import { ConfigType } from '@nestjs/config';
+import { lastValueFrom } from 'rxjs';
+
+import AcbsConfig from '../../config/acbs.config';
```
Comment (IgnatG) [inline src/modules/party/get-party-by-search-text.service.ts]: can we use the alias for this? i.e. `@ukef/config/...`

File: src/modules/party/get-party-by-search-text.service.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
+import { HttpService } from '@nestjs/axios';
```
Comment (IgnatG) [inline src/modules/party/get-party-by-search-text.service.ts]: could we not name this file as "party.service.ts`? or is it different?

File: src/modules/party/party.module.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
+import { HttpModule } from '@nestjs/axios';
+import { Module } from '@nestjs/common';
+import { ConfigModule, ConfigService } from '@nestjs/config';
+
+import { AcbsModule } from '../acbs/acbs.module';
```
Comment (IgnatG) [inline src/modules/party/party.module.ts]: should we use the alias for this path?

File: src/modules/party/dto/party-query.dto.ts (hunk 3 above)
Comment (oscar-richardson-softwire) [reply src/modules/party/dto/party-query.dto.ts]: Good point, you're right - I've made this change.

File: src/modules/party/get-party-by-search-text.service.ts (hunk 8 above)
Comment (lmarrai-sw) [reply src/modules/party/get-party-by-search-text.service.ts]: @oscar-richardson-softwire can you make sure we add a test that checks this logic as well please?

Comment (lmarrai-sw) [reply src/modules/party/get-party-by-search-text.service.ts]: Sorry, that message was a bit vague, let me try again.

As well as fixing this edge case where `OfficerRiskDate` is null, can you please also add a test that checks that we handle that particular case correctly?

**Comment by IgnatG** [comment]
> Should this endpoint use `searchtext` (as Mulesoft does) or `searchText` (in line with the camel casing used for other parameters in other endpoints, e.g. `partyIdentifier`)?

happy for it to change to `searchText`. Just remember to update the spreadsheet to indicate this change

**Comment by IgnatG** [comment]
> Should this endpoint be renamed to `/parties` in accordance with API spec guidelines, as it returns an array of parties that match the search text?

yes, please. remember to update the spreadsheet to indicate this change

File: src/modules/party/dto/party-query.dto.ts (hunk 4 above)
Comment (oscar-richardson-softwire) [reply src/modules/party/dto/party-query.dto.ts]: As we discussed on call, `@ApiProperty({minLength: 3})` adds that info to the OpenAPI yaml spec but it is not displayed on the Swagger UI. I've therefore added it as a description as well: `@ApiProperty({description: 'Minimum length: 3', minLength: 3})`. If the `minLength` property ends up being displayed when we start using Azure (and hence a different UI), we can just remove the `description` property.

File: test/party/get-party-by-search-text.api-test.ts (hunk 6 above)
Comment (oscar-richardson-softwire) [reply test/party/get-party-by-search-text.api-test.ts]: I've updated the function as follows:
```
public stringOfNumericCharacters(minLength?: number): string {
    
const stringOptions: Partial<Chance.StringOptions> = { pool: '0123456789' };

if (minLength) {      
const length = this.chance.integer({ min: minLength, max: Math.max(20, minLength*2)});      
stringOptions.length = length;
}    

return this.chance.string(stringOptions);  
}
```

File: test/party/get-party-by-search-text.api-test.ts (hunk 7 above)
Comment (oscar-richardson-softwire) [reply test/party/get-party-by-search-text.api-test.ts]: Good shout, added

File: src/modules/party/get-party-by-search-text.service.ts (hunk 8 above)
Comment (oscar-richardson-softwire) [reply src/modules/party/get-party-by-search-text.service.ts]: Added the logic and unit test for this

File: src/modules/party/party.controller.ts (hunk 10 above)
Comment (oscar-richardson-softwire) [reply src/modules/party/party.controller.ts]: I went with `getPartiesBySearchText`

File: src/modules/party/dto/get-party-by-search-text-response-element.dto.ts (hunk 11 above)
Comment (oscar-richardson-softwire) [reply src/modules/party/dto/get-party-by-search-text-response-element.dto.ts]: This is the class for the response body that we return from this endpoint. Do you want us to validate response bodies as well as requests?

File: src/modules/party/party.module.ts (hunk 14 above)
Comment (oscar-richardson-softwire) [reply src/modules/party/party.module.ts]: Changed to `'@ukef/module/acbs/acbs.module'`

File: src/modules/party/get-party-by-search-text.service.ts (hunk 13 above)
Comment (oscar-richardson-softwire) [reply src/modules/party/get-party-by-search-text.service.ts]: Good point, have renamed

**Comment by oscar-richardson-softwire** [comment]
> > Should this endpoint be renamed to `/parties` in accordance with API spec guidelines, as it returns an array of parties that match the search text?
> 
> yes, please. remember to update the spreadsheet to indicate this change

Cool, I've updated the spreadsheet

**Comment by oscar-richardson-softwire** [comment]
I also added a unit test for the case when there are no matching parties in ACBS (i.e., ACBS returns an empty array)

File: src/modules/party/dto/get-party-by-search-text-response-element.dto.ts (hunk 11 above)
Comment (IgnatG) [reply src/modules/party/dto/get-party-by-search-text-response-element.dto.ts]: depends. do we know what the existing functionality does? - i.e. if we pass to ACBS a payload that is not entirely complete, then does it work, or does it return an error?

Comment (oscar-richardson-softwire) [reply src/modules/party/dto/get-party-by-search-text-response-element.dto.ts]: Just to make sure we're on the same page, we **are** validating everything that we send to ACBS. The class we're looking at here is just the response shape we get from our endpoint

File: .env.test [hunk 15]
```diff
@@ -0,0 +1,19 @@
+# Login details and Dev config for TFS API
```
Comment (IgnatG) [inline .env.test:1]: I thought we are not using this file anymore?

Comment (oscar-richardson-softwire) [reply .env.test:1]: You're right we're not, I branched off an earlier version of the codebase so I'll need to rebase before I merge onto main
